agents/cluster_methods_mixin.py

```
@@ -1,188 +1,285 @@
+        """Compute per-language skip sets so the final combined cluster string fits."""
+        char_budget = self._cluster_prompt_budget(prompt_overhead_chars)
+        if char_budget <= 0:
+            ctx = get_current_agent_context_window()
+            msg = (
+                f"Prompt overhead ({prompt_overhead_chars} chars) consumes the entire agent input "
+                f"window ({_describe_window(ctx)}); no room for cluster renderings."
+            )
+            logger.error("[CFG skip planner] %s", msg)
+            raise ContextBudgetExceededError(msg, telemetry_properties=_window_telemetry(ctx, char_budget))
+
+        langs_with_clusters = [l for l in programming_langs if cluster_results.get(l)]
```

Ruff (0.16.0)

[error] Ambiguous variable name: `l`

(E741)